src/components/markdown-renderer/basic-markdown-renderer.tsx

```diff
@@ -4,43 +4,105 @@
+    onLineMarkerPositionChanged,
+    onFrontmatterChange,
+    onTaskCheckedChange,
+    onTocChange,
+    baseUrl,
+    onImageClick,
+    outerContainerRef
   }) => {
-  const maxLength = useSelector((state: ApplicationState) => state.config.maxDocumentLength)
-  const trimmedContent = useMemo(() => content.length > maxLength ? content.substr(0, maxLength) : content, [content,
-    maxLength])
-  const markdownReactDom = useConvertMarkdownToReactDom(trimmedContent, markdownIt, componentReplacers, onBeforeRendering, onAfterRendering)
+  const rawMetaRef = useRef<RawNoteFrontmatter>()
+  const markdownBodyRef = useRef<HTMLDivElement>(null)
+  const currentLineMarkers = useRef<LineMarkers[]>()
+  const hasNewYamlError = useRef(false)
+  const tocAst = useRef<TocAst>()
+  const [showYamlError, setShowYamlError] = useState(false)
+  const [trimmedContent, isTrimmed] = useTrimmedContent(content)
+
+  const markdownIt = useMemo(() =>
+    new BasicMarkdownItConfigurator({
+      useFrontmatter: !!onFrontmatterChange,
+      onParseError: errorState => hasNewYamlError.current = errorState,
+      onRawMetaChange: rawMeta => rawMetaRef.current = rawMeta,
+      onToc: toc => tocAst.current = toc,
+      onLineMarkers: onLineMarkerPositionChanged === undefined ? undefined
+        : lineMarkers => currentLineMarkers.current = lineMarkers
+    }).buildConfiguredMarkdownIt(), [onLineMarkerPositionChanged, onFrontmatterChange])
+
+  const clearFrontmatter = useCallback(() => {
+    hasNewYamlError.current = false
+    rawMetaRef.current = undefined
+    onBeforeRendering?.()
+  }, [onBeforeRendering])
+
+  const checkYamlErrorState = useCallback(() => {
+    setShowYamlError(hasNewYamlError.current)
+    onAfterRendering?.()
+  }, [onAfterRendering])
+
+  const baseReplacers = useComponentReplacers(onTaskCheckedChange, onImageClick, baseUrl)
+  const markdownReactDom = useConvertMarkdownToReactDom(trimmedContent, markdownIt, baseReplacers, additionalReplacers, clearFrontmatter, checkYamlErrorState)
+
+  useTranslation()
+  useCalculateLineMarkerPosition(markdownBodyRef, currentLineMarkers.current, onLineMarkerPositionChanged, markdownBodyRef.current?.offsetTop ?? 0)
+  useExtractFirstHeadline(markdownBodyRef, content, onFirstHeadingChange)
+  useOnRefChange(tocAst, onTocChange)
+  useOnRefChange(rawMetaRef, (newValue) => {
+    if (!newValue) {
+      onFrontmatterChange?.(undefined)
+    } else {
+      onFrontmatterChange?.(new NoteFrontmatter(newValue))
+    }
+  })
 
   return (
-    <div className={ `${ className ?? '' } d-flex flex-column align-items-center` }>
-      <DocumentLengthLimitReachedAlert contentLength={ content.length }/>
-      <div ref={ documentReference } className={ 'markdown-body w-100 d-flex flex-column align-items-center' }>
+    <div ref={ outerContainerRef } className={ 'position-relative' }>
+      <InvalidYamlAlert show={ showYamlError }/>
+      <DocumentLengthLimitReachedAlert show={ isTrimmed }/>
```

Comment:
The variable name `isTrimmed` is a bit misleading here, as one could see isTrimmed as the note is already short. Maybe rename it to `exceededLimit`?

Reply:
Renamed